## Predizendo preços de casas

Queremos pegar dados sobre diversas casas e tentar advinhar quais os preços que elas serão vendidas. Para isso vamos usar um modelo de aprendizagem de máquina que simule essa situação.

Clicar com o botão direito e mandar um save link as: https://github.com/cassiasamp/semcomp/tree/master/dados

Precisamos ler os dados, para ler os dados vamos usar o pandas.

In [0]:
import pandas as pd

In [0]:
dados_treino = pd.read_csv('treino.csv')

Agora que fizemos o carregamento dos dados em um dataframe do pandas, podemos dar uma olhada neles com .head().

In [3]:
dados_treino.head(5)

Id  MSSubClass MSZoning  ...  SaleType  SaleCondition SalePrice
0   1          60       RL  ...        WD         Normal    208500
1   2          20       RL  ...        WD         Normal    181500
2   3          60       RL  ...        WD         Normal    223500
3   4          70       RL  ...        WD        Abnorml    140000
4   5          60       RL  ...        WD         Normal    250000

[5 rows x 81 columns]

Repare que ficamos com uma coluna com 0, 1, 2, 3, 4 e que já temos nos dados uma coluna chamada Id, então vamos arrumar isso dizendo para o pandas que essa coluna já é um índice.

In [0]:
dados_treino = pd.read_csv('treino.csv', index_col='Id')

Vamos dar uma olhada novamente para ver se está tudo certo com os dados.

In [8]:
dados_treino.head(5)

MSSubClass MSZoning  LotFrontage  ...  SaleType SaleCondition SalePrice
Id                                    ...                                  
1           60       RL         65.0  ...        WD        Normal    208500
2           20       RL         80.0  ...        WD        Normal    181500
3           60       RL         68.0  ...        WD        Normal    223500
4           70       RL         60.0  ...        WD       Abnorml    140000
5           60       RL         84.0  ...        WD        Normal    250000

[5 rows x 80 columns]

In [24]:
dados_treino.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1460 entries, 1 to 1460
Data columns (total 80 columns):
MSSubClass       1460 non-null int64
MSZoning         1460 non-null object
LotFrontage      1201 non-null float64
LotArea          1460 non-null int64
Street           1460 non-null object
Alley            91 non-null object
LotShape         1460 non-null object
LandContour      1460 non-null object
Utilities        1460 non-null object
LotConfig        1460 non-null object
LandSlope        1460 non-null object
Neighborhood     1460 non-null object
Condition1       1460 non-null object
Condition2       1460 non-null object
BldgType         1460 non-null object
HouseStyle       1460 non-null object
OverallQual      1460 non-null int64
OverallCond      1460 non-null int64
YearBuilt        1460 non-null int64
YearRemodAdd     1460 non-null int64
RoofStyle        1460 non-null object
RoofMatl         1460 non-null object
Exterior1st      1460 non-null object
Exterior2nd      1460 non-

Repare que temos 80 colunas nos nossos dados, e 1460 entradas.

Legal, temos os dados de treino, agora falta importarmos os dados de teste.

In [0]:
dados_teste = pd.read_csv('teste.csv')

Vamos dar uma olhada nos dados também.

In [11]:
dados_teste.head(5)

Id  MSSubClass MSZoning  ...  YrSold  SaleType SaleCondition
0  1461          20       RH  ...    2010        WD        Normal
1  1462          20       RL  ...    2010        WD        Normal
2  1463          60       RL  ...    2010        WD        Normal
3  1464          60       RL  ...    2010        WD        Normal
4  1465         120       RL  ...    2010        WD        Normal

[5 rows x 80 columns]

Repare que temos a mesma situação com o id, então, vamos dizer qual a coluna de indice.

In [0]:
dados_teste = pd.read_csv('teste.csv', index_col='Id')

Vamos conferir se está tudo certo novamente.

In [13]:
dados_teste.head(5)

MSSubClass MSZoning  LotFrontage  ...  YrSold SaleType SaleCondition
Id                                      ...                               
1461          20       RH         80.0  ...    2010       WD        Normal
1462          20       RL         81.0  ...    2010       WD        Normal
1463          60       RL         74.0  ...    2010       WD        Normal
1464          60       RL         78.0  ...    2010       WD        Normal
1465         120       RL         43.0  ...    2010       WD        Normal

[5 rows x 79 columns]

Beleza, agora temos nossos dados de treino e teste, podemos começar a mexer neles. Vamos dar uma olhada nas colunas dos nossos dados de treino.

In [14]:
dados_treino.columns

Index(['MSSubClass', 'MSZoning', 'LotFrontage', 'LotArea', 'Street', 'Alley',
       'LotShape', 'LandContour', 'Utilities', 'LotConfig', 'LandSlope',
       'Neighborhood', 'Condition1', 'Condition2', 'BldgType', 'HouseStyle',
       'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd', 'RoofStyle',
       'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType', 'MasVnrArea',
       'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual', 'BsmtCond',
       'BsmtExposure', 'BsmtFinType1', 'BsmtFinSF1', 'BsmtFinType2',
       'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'Heating', 'HeatingQC',
       'CentralAir', 'Electrical', '1stFlrSF', '2ndFlrSF', 'LowQualFinSF',
       'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath', 'HalfBath',
       'BedroomAbvGr', 'KitchenAbvGr', 'KitchenQual', 'TotRmsAbvGrd',
       'Functional', 'Fireplaces', 'FireplaceQu', 'GarageType', 'GarageYrBlt',
       'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual', 'GarageCond',
       'PavedDrive', 'Wo

Repare naquela última coluna, ela é coluna que queremos advinhar pois tem os preços de venda das casas e é chamada **SalePrice**.
Como SalePrice é o que queremos advinhar, vou criar um novo dataframe que contém apenas essa coluna e chamá-lo de y.

In [0]:
y = dados_treino.SalePrice

Vamos dar uma olhadinha no nosso y.

In [16]:
y.head(5)

Id
1    208500
2    181500
3    223500
4    140000
5    250000
Name: SalePrice, dtype: int64

Repare que temos o id e os preços de cada casa. Tudo certo.

Temos o nosso y, e qual será o nosso X? Podemos escolher algumas características dos nossos dados, algumas dessas colunas para fazermos a predição.
Vou escolher 7 colunas:


*   área do lote ou **LotArea**
*   ano de construção da casa ou **YearBuilt**
*   tamanho do 1° andar ou **1stFlrSF**
*   tamanho do 2° andar ou **2ndFlrSF**
*   banheiros completos acima do térreo ou **FullBath**
*   dormitórios acima do térreo  ou **BedroomAbvGr**
*   total de quartos acima do térreo ou **TotRmsAbvGrd**



In [0]:
caracteristicas = ['LotArea', 'YearBuilt', '1stFlrSF', '2ndFlrSF', 
                   'FullBath', 'BedroomAbvGr','TotRmsAbvGrd']

Para garantir que não estamos alterando os nossos dados de treino originais, vamos criar uma cópia deles para ser o nosso X apenas com as características que queremos usar.

Fazemos isso usando o .copy() do pandas.

In [0]:
X = dados_treino[caracteristicas].copy()

In [22]:
X.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1460 entries, 1 to 1460
Data columns (total 7 columns):
LotArea         1460 non-null int64
YearBuilt       1460 non-null int64
1stFlrSF        1460 non-null int64
2ndFlrSF        1460 non-null int64
FullBath        1460 non-null int64
BedroomAbvGr    1460 non-null int64
TotRmsAbvGrd    1460 non-null int64
dtypes: int64(7)
memory usage: 91.2 KB


Beleza, agora temos nosso X e y, o que podemos fazer? Dividir os nossos dados para alimentar nosso modelo.

Para evitarmos usar os nossos dados de teste nesse momento, vamos dividir os dados de X e y entre treino e validação. Usaremos os dados de teste mais para frente apenas para saber como o nosso modelo está generalizando.

Nessa divisão vou reservar 80% dos dados para treino e 20% para validação. Aqui também não há uma regra.

Assim como chamar de X e y, há diferentes formar de nomear as divisões no split, por exemplo, podemos usar treino e teste aqui e validacao para o que vamos fazer com teste.


Para isso precisamos importar a função de split do sklearn e vamos definir um estados chamado random_state como 42 para garantir que o resultado permaneça o mesmo.

In [0]:
from sklearn.model_selection import train_test_split

X_treino, X_valid, y_treino, y_valid = train_test_split(X, y, 
                                                    train_size=0.8, 
                                                    test_size=0.2, 
                                                    random_state=42)

Agora que dividimos os nossos dados, podemos criar um modelo e alimentá-lo ou fazer o fit. Vamos usar um modelo de árvore de decisão chamado DecisionTreeRegressor.

Vamos adicionar aqui também um random_state para garantir que os resultados sejam os mesmos ou sua replicabilidade.

In [0]:
from sklearn import tree

modelo = tree.DecisionTreeRegressor(random_state=42)

In [33]:
modelo.fit(X, y)

DecisionTreeRegressor(criterion='mse', max_depth=None, max_features=None,
                      max_leaf_nodes=None, min_impurity_decrease=0.0,
                      min_impurity_split=None, min_samples_leaf=1,
                      min_samples_split=2, min_weight_fraction_leaf=0.0,
                      presort=False, random_state=42, splitter='best')

Depois do fit o que falta? Fazer predicoes, tentar adivinhar os preços das casas dos dados. Então, vamos lá fazer isso com .predict() e os dados de validação.

In [0]:
predicoes = modelo.predict(X_valid)

Bacana, acabamos de ajustar o nosso modelo de árvore com ao nosso X e y e fazer a predição, mas será que ele é bom? Para saber como ele está indo, vamos usar uma medida de erro chamada MAE.

In [38]:
from sklearn.metrics import mean_absolute_error

mae = mean_absolute_error(y_valid, predicoes)

mae

86.82534246575342

E veja o que a nossa métrica nos devolveu, estamos errando com 86,8 doláres de diferença no geral. Podemos mostrar isso de um jeito mais legal, inclusive.

In [41]:
print('MAE do modelo de árvore: %d'% mae)

MAE do modelo de árvore: 86


Nós pegamos os nossos dados, fizemos cópias, escolhemos features, separamos, criamos o modelo, ajustamos, predizemos, vimos a métrica e agora? Agora falta saber como seria no teste, como seria no mundo real, para isso, vamos usar o nosso conjunto de dados de teste.

Como selecionamos 7 caracteristicas ou features para os dados de treino, precisamos fazer o mesmo para os dados de teste, senão não temos como comparar um com o outro, então vamos repetir a selecao de features e a copia do dataframe.

In [0]:
X_teste = dados_teste[caracteristicas].copy()

In [28]:
X_teste.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1459 entries, 1461 to 2919
Data columns (total 7 columns):
LotArea         1459 non-null int64
YearBuilt       1459 non-null int64
1stFlrSF        1459 non-null int64
2ndFlrSF        1459 non-null int64
FullBath        1459 non-null int64
BedroomAbvGr    1459 non-null int64
TotRmsAbvGrd    1459 non-null int64
dtypes: int64(7)
memory usage: 91.2 KB


In [0]:
predicoes_teste = modelo.predict(X_teste)

Beleza, agora fizemos as predicoes para dados desconhecidos e o que podemos fazer? Podemos salvar esses resultado em um arquivo.

In [49]:
predicoes_teste

array([ 94750., 137500., 176000., ..., 157900.,  93500., 220000.])

Repare que o que as predicoes nos devolvem são arrays, entao podemos colocar esses valores em um dicionario identificando os campos. Para cada entrada de uma casa no indice, temos um valor de predicao.

In [0]:
resultados_arvore = {'Id': X_teste.index, 'SalePrice': predicoes_teste}

Agora, podemos passar esse dicionario para um dataframe e retornar um csv.

Do mesmo jeito que carregamos os nossos dados com read_csv, podemos usar o pandas para salvar os nossos dados nesse formato com to_csv. 

Quando o pandas faz isso ele cria automaticamente uma coluna de índice, mas queremos manter o Id como indice, então vamos passar um index=False.

In [0]:
resultados = pd.DataFrame(resultados_arvore)
resultados.to_csv('resultados_arvore.csv', index=False)